ext/websocket/server.rs

```diff
@@ -0,0 +1,191 @@
+  state: Rc<RefCell<OpState>>,
+  rid: ResourceId,
+  data: ZeroCopyBuf,
+) -> Result<(), AnyError> {
+  let resource = state
+    .borrow_mut()
+    .resource_table
+    .get::<ServerWebSocket>(rid)?;
+
+  let mut ws = RcRef::map(&resource, |r| &r.ws).borrow_mut().await;
+  ws.write_frame(Frame::new(true, OpCode::Binary, None, data.to_vec()))
+    .await
+    .map_err(|err| type_error(err.to_string()))?;
+  Ok(())
+}
+
+#[op]
+pub async fn op_server_ws_send_text(
+  state: Rc<RefCell<OpState>>,
+  rid: ResourceId,
+  data: String,
+) -> Result<(), AnyError> {
+  let resource = state
+    .borrow_mut()
+    .resource_table
+    .get::<ServerWebSocket>(rid)?;
+  let mut ws = RcRef::map(&resource, |r| &r.ws).borrow_mut().await;
+  ws.write_frame(Frame::new(true, OpCode::Text, None, data.into_bytes()))
+    .await
+    .map_err(|err| type_error(err.to_string()))?;
+  Ok(())
+}
+
+#[op]
+pub async fn op_server_ws_send(
+  state: Rc<RefCell<OpState>>,
+  rid: ResourceId,
+  value: SendValue,
+) -> Result<(), AnyError> {
+  let msg = match value {
+    SendValue::Text(text) => {
+      Frame::new(true, OpCode::Text, None, text.into_bytes())
+    }
+    SendValue::Binary(buf) => {
+      Frame::new(true, OpCode::Binary, None, buf.to_vec())
+    }
+    SendValue::Pong => Frame::new(true, OpCode::Pong, None, vec![]),
+    SendValue::Ping => Frame::new(true, OpCode::Ping, None, vec![]),
+  };
+
+  let resource = state
+    .borrow_mut()
+    .resource_table
+    .get::<ServerWebSocket>(rid)?;
+  let mut ws = RcRef::map(&resource, |r| &r.ws).borrow_mut().await;
+
+  ws.write_frame(msg)
+    .await
+    .map_err(|err| type_error(err.to_string()))?;
+  Ok(())
+}
+
+#[op(deferred)]
+pub async fn op_server_ws_close(
+  state: Rc<RefCell<OpState>>,
+  rid: ResourceId,
+  code: Option<u16>,
+  reason: Option<String>,
+) -> Result<(), AnyError> {
+  let resource = state
+    .borrow_mut()
+    .resource_table
+    .get::<ServerWebSocket>(rid)?;
+  let mut ws = RcRef::map(&resource, |r| &r.ws).borrow_mut().await;
+  let frame = reason
+    .map(|reason| Frame::close(code.unwrap_or(1005), reason.as_bytes()))
+    .unwrap_or_else(|| Frame::close_raw(vec![]));
+  ws.write_frame(frame)
+    .await
+    .map_err(|err| type_error(err.to_string()))?;
+  Ok(())
+}
+
+#[op]
+pub async fn op_server_ws_next_event(
+  state: Rc<RefCell<OpState>>,
+  rid: ResourceId,
+) -> Result<(u16, StringOrBuffer), AnyError> {
+  let resource = state
+    .borrow_mut()
+    .resource_table
+    .get::<ServerWebSocket>(rid)?;
+  let mut ws = RcRef::map(&resource, |r| &r.ws).borrow_mut().await;
+  let val = match ws.read_frame().await {
+    Ok(val) => val,
+    Err(err) => {
+      return Ok((
+        MessageKind::Error as u16,
+        StringOrBuffer::String(err.to_string()),
+      ))
+    }
+  };
+
+  let res = match val.opcode {
+    OpCode::Text => (
+      MessageKind::Text as u16,
+      StringOrBuffer::String(String::from_utf8(val.payload).unwrap()),
```

Comment:
Should this return an error instead of `panic`ing?

Reply:
fastws guarantees Text payloads are always utf8. Its not obvious yet in the documentation, when thats done we can switch this to `String::from_utf8_unchecked`